core/shared/src/main/scala/org/http4s/StaticFile.scala

```diff
@@ -86,57 +95,74 @@ object StaticFile {
-
-        if (expired) {
-          val len = urlConn.getContentLengthLong
-          val headers = Headers(
-            lastmod.map(`Last-Modified`(_)),
-            nameToContentType(url.getPath),
-            if (len >= 0) `Content-Length`.unsafeFromLong(len)
-            else `Transfer-Encoding`(TransferCoding.chunked.pure[NonEmptyList]),
-          )
-
-          F.blocking(urlConn.getInputStream)
-            .redeem(
-              recover = {
-                case _: FileNotFoundException => None
-                case other => throw other
-              },
-              f = { inputStream =>
-                Some(
-                  Response(
-                    headers = headers,
-                    body = readInputStream[F](F.pure(inputStream), DefaultBufferSize),
+        F.blocking(url.openConnection).flatMap { urlConn =>
+          (F.blocking(urlConn.getLastModified()), F.blocking(urlConn.getContentLengthLong())).tupled
+            .flatMap { case (lastModified, contentLength) =>
+              val lastmod = HttpDate.fromEpochSecond(lastModified / 1000).toOption
+              val ifModifiedSince: Option[`If-Modified-Since`] =
+                req.flatMap(_.headers.get[`If-Modified-Since`])
+              val expired = (ifModifiedSince, lastmod).mapN(_.date < _).getOrElse(true)
+
+              if (expired) {
+                etagCalculator(url).flatMap { etag =>
+                  val headers = Headers(
+                    lastmod.map(`Last-Modified`(_)),
+                    nameToContentType(url.getPath),
+                    etag,
+                    if (contentLength >= 0) `Content-Length`.unsafeFromLong(contentLength)
+                    else `Transfer-Encoding`(TransferCoding.chunked.pure[NonEmptyList]),
                   )
-                )
-              },
-            )
-        } else
-          F.blocking(urlConn.getInputStream.close())
-            .handleError(_ => ())
-            .as(Some(Response(NotModified)))
+                  F.blocking(urlConn.getInputStream)
+                    .redeem(
+                      recover = {
+                        case _: FileNotFoundException => None
+                        case other => throw other
+                      },
+                      f = { inputStream =>
+                        Some(
+                          Response(
+                            headers = headers,
+                            body = readInputStream[F](F.pure(inputStream), DefaultBufferSize),
+                          )
+                        )
+                      },
+                    )
+                }
+              } else
+                F.blocking(urlConn.getInputStream.close())
+                  .handleError(_ => ())
+                  .as(Some(Response(NotModified)))
+            }
+        }
       }
     })
   }
 
+  private def calcETagURL[F[_]](implicit F: Sync[F]): URL => F[Option[ETag]] = url =>
+    for {
+      urlConn <- F.blocking(url.openConnection)
```

Comment:
Is it weird that the connection is being opened twice? Both here, and in `fromURL`. Maybe it's not that big of a deal.

Reply:
I don't know.  I thought about just passing contentLength/lastModified time directly to the etag function instead of URL but if they wanted to do more (hash the actual contents) that wouldn't be that helpful.  We could pass the actual `URLConnection` but not sure that improves much.

> Returns a [URLConnection](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html) instance that represents a connection to the remote object referred to by the URL.

> A new instance of [URLConnection](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html) is created every time when invoking the [URLStreamHandler.openConnection(URL)](https://docs.oracle.com/javase/7/docs/api/java/net/URLStreamHandler.html#openConnection(java.net.URL)) method of the protocol handler for this URL.

> It should be noted that a URLConnection instance does not establish the actual network connection on creation. This will happen only when calling [URLConnection.connect()](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#connect()).

I dug a little unto the gnarly, mutable, impure mess of JarURLConnection and can't find anything too bad.  We could change etagCalculator to take some sort of `UrlContext` trait:

```scala
trait UrlContext {
  def contentLength: Option[Long]
  def lastModified: Option[Long]
  def urlConnection: UrlConnection
  def url: Url
}
```

This would let the default implentation reuse connection and contentLength/lastModified without re-opening the connection.  By making it a trait it allows us to add members in the future without breaking API.  Might be a bit over-engineering though.

Reply:
The fact that `openConnection` may throw an `IOException` suggests to me that it may be doing something non-trivial. So maybe the etag calculator should be in terms of `URLConnection` so it can be opened once and shared.

Would be great to get another opinion on this.

Reply:
@armanbilge So one option that could move this PR forward is we just remove the ability to customize the etag.  Do we really need to support that?  If we remove that, the internal, canonical implementation can just use the opened connection, parameters listed above.  Seems to me the only thing keeping this moving forward is cause we don't want to commit to an API change that we would be locked into.

Reply:
@matthughes yes, that seems like a good solution to me :) we can always look into adding this API in the future if someone wants to customize it.

Reply:
Hey! What's the status of this PR? Let me know if I can help somehow : )

Reply:
@diogocanut if you want to open a new PR based on this branch and making the changes discussed here. I think we decided to hide this API in the end?

Btw thanks for all your work on Typelevel projects lately, you've been awesome 🚀

Reply:
Moved it to #7155 . I implement this method in terms of `URLConnection` so we don't have to use `openConnection` twice, it was the final decision here right?